src/eth.rs

```diff
@@ -1,75 +1,163 @@
+    signer: Option<SignerMiddleware<Provider<Http>, LocalWallet>>,
+    /// A Deployed Solidity Contract Address. This is required to interact with the Banyan Contract.
+    contract: Option<Contract<Provider<Http>>>,
+    /// A timeout for the Eth Client
+    timeout: Duration,
 }
 
-impl VitalikProvider {
-    pub fn new(url: String, contract_address: String) -> Result<Self> {
-        let provider = Provider::<Http>::try_from(url)?;
-        let provider2 = provider.clone();
-        let address = contract_address
-            .parse::<Address>()
-            .expect("could not parse contract address");
-        let abi: Abi = serde_json::from_str(
-            fs::read_to_string("contract_abi.json")
-                .expect("can't read file")
-                .as_str(),
-        )
-        .expect("couldn't load abi");
-
-        // TODO is this the right place to be sticking mutexes?
-        Ok(Self {
-            provider: Mutex::new(provider),
-            // my_pending_transactions: HashMap::new(),
-            //timeout: Duration::from_secs(timeout_seconds),
-            contract: Mutex::new(Contract::new(address, abi, provider2)),
-        })
+/// EthClientBuilder - A Default Builder for the EthClient
+impl Default for EthClient {
+    fn default() -> Self {
+        EthClientBuilder::default().build().unwrap()
+    }
+}
+
+#[derive(Clone, Debug, Copy, EthEvent)]
+struct NewOffer {
+    // #[indexed]
+    pub creator: Address,
+    // #[indexed]
+    pub executor: Address,
+    // #[indexed]
+    pub offerId: U256, // TODO: Snake case
+}
+
+/// The EthProvider is a wrapper around the ethers-rs Provider that handles all Ethereum interactions.
+impl EthClient {
+    /// Create a new default EthClientBuilder
+    pub fn builder() -> EthClientBuilder {
+        EthClientBuilder::default()
+    }
+
+    /* Struct State Methods */
+
+    /// Return whether theres's a signer configured
+    pub fn has_signer(&self) -> bool {
+        self.signer.is_some()
+    }
+
+    /// Return whether theres's a contract configured
+    pub fn has_contract(&self) -> bool {
+        self.contract.is_some()
+    }
+
+    /* Banyan Functions */
+
+    /* Deal Stuff */
+
+    #[allow(unused_variables)]
+    pub async fn propose_deal(&self, dp: DealProposal) -> Result<DealID, Error> {
+        if !self.has_signer() {
+            return Err(anyhow!("No signer available"));
+        }
+        if !self.has_contract() {
+            return Err(anyhow!("No contract available"));
+        }
+        // Borrow our signer and contract
+        let signer = self.signer.as_ref().unwrap();
+        let contract = self.contract.as_ref().unwrap();
+        // TODO: Tokenize the DealProposal
+        dbg!("DP: {:?}", &dp);
+
+        dbg!("Tokenizing Arguments");
+        let data = contract.encode("startOffer", dp).unwrap();
+        // TODO: Create a transaction request
+        dbg!("Initializing mew Request");
```

Comment:
typo

Reply:
wanna complete that pokedex tho. (fixed)